src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/4.diff.sql

```diff
@@ -83,5 +83,229 @@ AS
+    @sopInstanceUid         VARCHAR(64),               
+    @stringExtendedQueryTags dbo.InsertStringExtendedQueryTagTableType_1 READONLY,    
+    @longExtendedQueryTags dbo.InsertLongExtendedQueryTagTableType_1 READONLY,
+    @doubleExtendedQueryTags dbo.InsertDoubleExtendedQueryTagTableType_1 READONLY,
+    @dateTimeExtendedQueryTags dbo.InsertDateTimeExtendedQueryTagTableType_1 READONLY,
+    @personNameExtendedQueryTags dbo.InsertPersonNameExtendedQueryTagTableType_1 READONLY
+AS
+    SET NOCOUNT    ON
+    SET XACT_ABORT ON
+    BEGIN TRANSACTION
+
+        DECLARE @studyKey BIGINT
+        DECLARE @seriesKey BIGINT
+        DECLARE @instanceKey BIGINT
+        DECLARE @watermark BIGINT
+
+        -- Add lock so that the instance won't be removed
+        DECLARE @status TINYINT
+        SELECT
+            @studyKey = StudyKey,
+            @seriesKey = SeriesKey,
+            @instanceKey = InstanceKey,
+            @watermark = Watermark,
+            @status = Status
+        FROM dbo.Instance WITH (HOLDLOCK) 
+        WHERE StudyInstanceUid = @studyInstanceUid
+            AND SeriesInstanceUid = @seriesInstanceUid
+            AND SopInstanceUid = @sopInstanceUid
+
+        IF @@ROWCOUNT = 0
+            THROW 50404, 'Instance not exists', 1
+        IF @status <> 1 -- Created
+            THROW 50409, 'Instance is not been stored succssfully', 1
+
+        -- Insert Extended Query Tags
+
+        -- String Key tags
+        IF EXISTS (SELECT 1 FROM @stringExtendedQueryTags)
+        BEGIN      
+            MERGE INTO dbo.ExtendedQueryTagString AS T
+            USING 
+            (
+                -- Locks tags in dbo.ExtendedQueryTag
+                SELECT input.TagKey, input.TagValue, input.TagLevel 
+                FROM @stringExtendedQueryTags input
+                INNER JOIN dbo.ExtendedQueryTag WITH (REPEATABLEREAD) 
+                ON dbo.ExtendedQueryTag.TagKey = input.TagKey
+                -- Not merge on extended query tag which is being deleted.
+                AND dbo.ExtendedQueryTag.TagStatus <> 2     
+            ) AS S
+            ON T.TagKey = S.TagKey        
+                AND T.StudyKey = @studyKey
+                -- Null SeriesKey indicates a Study level tag, no need to compare SeriesKey
+                AND ISNULL(T.SeriesKey, @seriesKey) = @seriesKey      
+                -- Null InstanceKey indicates a Study/Series level tag, no to compare InstanceKey
+                AND ISNULL(T.InstanceKey, @instanceKey) = @instanceKey
+            WHEN MATCHED THEN
+                -- When index already exist, update only when watermark is newer
+                UPDATE SET T.Watermark = IIF(@watermark > T.Watermark, @watermark, T.Watermark), T.TagValue = IIF(@watermark > T.Watermark, S.TagValue, T.TagValue)
+            WHEN NOT MATCHED THEN 
+                INSERT (TagKey, TagValue, StudyKey, SeriesKey, InstanceKey, Watermark)
+                VALUES(
+                S.TagKey,
```

Comment:
nit: we should probably indent the values

Reply:
formatted. I think we do need a SQL formatter. :(

Reply:
lol yeah